chain/migrate.go

```diff
@@ -0,0 +1,159 @@
+package chain
+
+import (
+	"errors"
+	"fmt"
+
+	"go.sia.tech/core/consensus"
+	"go.sia.tech/core/types"
+)
+
+type oldSiacoinElement types.SiacoinElement
+
+func (oldSiacoinElement) Cast() (sce types.SiacoinElement) { return }
+
+func (sce *oldSiacoinElement) DecodeFrom(d *types.Decoder) {
+	sce.ID.DecodeFrom(d)
+	sce.StateElement.DecodeFrom(d)
+	(*types.V2SiacoinOutput)(&sce.SiacoinOutput).DecodeFrom(d)
+	sce.MaturityHeight = d.ReadUint64()
+}
+
+type oldSiafundElement types.SiafundElement
+
+func (oldSiafundElement) Cast() (sfe types.SiafundElement) { return }
+
+func (sfe *oldSiafundElement) DecodeFrom(d *types.Decoder) {
+	sfe.ID.DecodeFrom(d)
+	sfe.StateElement.DecodeFrom(d)
+	(*types.V2SiafundOutput)(&sfe.SiafundOutput).DecodeFrom(d)
+	(*types.V2Currency)(&sfe.ClaimStart).DecodeFrom(d)
+}
+
+type oldFileContractElement types.FileContractElement
+
+func (oldFileContractElement) Cast() (fce types.FileContractElement) { return }
+
+func (fce *oldFileContractElement) DecodeFrom(d *types.Decoder) {
+	fce.ID.DecodeFrom(d)
+	fce.StateElement.DecodeFrom(d)
+	fce.FileContract.DecodeFrom(d)
+}
+
+type oldTransactionSupplement consensus.V1TransactionSupplement
+
+func (oldTransactionSupplement) Cast() (ts consensus.V1TransactionSupplement) { return }
+
+func (ts *oldTransactionSupplement) DecodeFrom(d *types.Decoder) {
+	types.DecodeSliceCast[oldSiacoinElement](d, &ts.SiacoinInputs)
+	types.DecodeSliceCast[oldSiafundElement](d, &ts.SiafundInputs)
+	types.DecodeSliceCast[oldFileContractElement](d, &ts.RevisedFileContracts)
+	types.DecodeSliceFn(d, &ts.StorageProofs, func(d *types.Decoder) (sp consensus.V1StorageProofSupplement) {
+		(*oldFileContractElement)(&sp.FileContract).DecodeFrom(d)
+		return
+	})
+}
+
+type oldBlockSupplement consensus.V1BlockSupplement
+
+func (oldBlockSupplement) Cast() (bs consensus.V1BlockSupplement) { return }
+
+func (bs *oldBlockSupplement) DecodeFrom(d *types.Decoder) {
+	types.DecodeSliceCast[oldTransactionSupplement](d, &bs.Transactions)
+	types.DecodeSliceCast[oldFileContractElement](d, &bs.ExpiringFileContracts)
+}
+
+type oldSupplementedBlock supplementedBlock
+
+func (sb *oldSupplementedBlock) DecodeFrom(d *types.Decoder) {
+	if v := d.ReadUint8(); v != 2 {
+		d.SetErr(fmt.Errorf("incompatible version (%d)", v))
+	}
+	var b types.Block
+	(*types.V2Block)(&b).DecodeFrom(d)
+	sb.Block = &b
+	sb.Header = b.Header()
+	types.DecodePtrCast[oldBlockSupplement](d, &sb.Supplement)
+}
+
+type versionedState consensus.State
+
+func (vs *versionedState) DecodeFrom(d *types.Decoder) {
+	if v := d.ReadUint8(); v != 2 {
+		d.SetErr(fmt.Errorf("incompatible version (%d)", v))
+	}
+	(*consensus.State)(vs).DecodeFrom(d)
+}
+
+// MigrateDB upgrades the database to the latest version.
+func MigrateDB(db DB, n *consensus.Network) error {
+	if db.Bucket(bVersion) == nil {
+		return nil // nothing to migrate
```

Comment:
shouldn't we try to migrate from "no version" to version 2? rather than never migrating?

Reply:
turns out we've always had a version lol, I just forgot https://github.com/SiaFoundation/coreutils/blob/a28e868d07995cc1a917b3bd75e6645653c75901/chain/db.go#L780